### PR TITLE
Undo/redo for binary editor

### PR DESCRIPTION
**Pull request by hajdam**
Added undo/redo for binary editor

### CODE REVIEW

File: mucommander-viewer-binary/src/main/java/com/mucommander/viewer/binary/BinaryEditor.java [hunk 1]
```diff
@@ -139,8 +154,63 @@ private void initMenuBars() {
...
 
+        undoMenuItem = MenuToolkit.createMenuItem(
+                Translator.get("binary_editor.undo"),
+                menuItemMnemonicHelper,
+                DesktopManager.getActionShortcuts().getDefaultKeystroke(ActionType.Undo),
+                e -> {
+                    try {
+                        undoHandler.performUndo();
+                    } catch (BinaryDataOperationException ex) {
+                        Logger.getLogger(BinaryEditor.class.getName())
```
Comment (pskowronek) [inline mucommander-viewer-binary/src/main/java/com/mucommander/viewer/binary/BinaryEditor.java:165]: Make it a final member of BinaryEditor?

File: mucommander-viewer-binary/src/main/java/com/mucommander/viewer/binary/BinaryEditor.java [hunk 2]
```diff
@@ -139,8 +154,63 @@ private void initMenuBars() {
         editMenu.add(pasteMenuItem, copyMenuItemPosition + 2);
         editMenu.add(deleteMenuItem, copyMenuItemPosition + 3);
 
+        undoMenuItem = MenuToolkit.createMenuItem(
+                Translator.get("binary_editor.undo"),
+                menuItemMnemonicHelper,
+                DesktopManager.getActionShortcuts().getDefaultKeystroke(ActionType.Undo),
+                e -> {
+                    try {
```
Comment (pskowronek) [inline mucommander-viewer-binary/src/main/java/com/mucommander/viewer/binary/BinaryEditor.java:162]: Move the body to a new method that catches and logs the exception. This way you could simply do sth like this: `e -> doUndo` - a cleaner code, plus you would avoid code duplication (there's another occurrence of undo code below).

File: mucommander-viewer-binary/src/main/java/com/mucommander/viewer/binary/BinaryEditor.java [hunk 3]
```diff
@@ -139,8 +154,63 @@ private void initMenuBars() {
...
+                    } catch (BinaryDataOperationException ex) {
+                        Logger.getLogger(BinaryEditor.class.getName())
+                                .log(Level.FINE, "Undo operation failed", ex);
+                    }
+                });
+        redoMenuItem = MenuToolkit.createMenuItem(
+                Translator.get("binary_editor.redo"),
+                menuItemMnemonicHelper,
+                DesktopManager.getActionShortcuts().getDefaultKeystroke(ActionType.Redo),
+                e -> {
```
Comment (pskowronek) [inline mucommander-viewer-binary/src/main/java/com/mucommander/viewer/binary/BinaryEditor.java:173]: the same for Redo

**Review by pskowronek — CHANGES_REQUESTED** [review]
see my comments

**Comment by pskowronek** [comment]
Thanks for updates @hajdam, no more remarks from my side. @ahadas ?

**Review by pskowronek — APPROVED** [review]
looks good from my side.

File: mucommander-viewer-binary/src/main/java/com/mucommander/viewer/binary/BinaryEditor.java [hunk 4]
```diff
@@ -377,4 +434,23 @@ public void setPresenter(EditorPresenter presenter) {
...
+
+    public enum BinaryEditorAction implements DialogAction {
+        YES("save"),
+        NO("dont_save"),
+        CANCEL("cancel");
+
+        private final String actionName;
+
+        BinaryEditorAction(@Nonnull String actionKey) {
+            // here or when in #getActionName
```
Comment (ahadas) [inline mucommander-viewer-binary/src/main/java/com/mucommander/viewer/binary/BinaryEditor.java:446]: it's ok here, this comment can be dropped

**Review by ahadas — COMMENTED** [review]
@hajdam thanks, besides the minor comment inside can you please add a line about this in the [readme file](https://github.com/mucommander/mucommander/blob/master/package/readme.txt)?

**Comment by pskowronek** [comment]
> @hajdam thanks, besides the minor comment inside can you please add a line about this in the [readme file](https://github.com/mucommander/mucommander/blob/master/package/readme.txt)?

@hajdam the changes are good, but I guess @ahadas meant to add a note to either `New featues` or `Improvements' section?

**Comment by hajdam** [comment]
> > @hajdam thanks, besides the minor comment inside can you please add a line about this in the [readme file](https://github.com/mucommander/mucommander/blob/master/package/readme.txt)?
> 
> @hajdam the changes are good, but I guess @ahadas meant to add a note to either `New featues` or `Improvements` section?

Oh, I didn't notice that section :-|

**Review by ahadas — APPROVED** [review]
thanks!

**Comment by pskowronek** [comment]
@hajdam  any chance for Find/Search support in Binary Editor? :)

**Comment by hajdam** [comment]
> @hajdam any chance for Find/Search support in Binary Editor? :)

I could copy find/search capability from standalone app, but as I understand it, binary viewer module should be kept in minimal form, because internal viewers/editors are not part of the core functionality in the muCommander.
Not sure how far it is allowed to be expanded... :-|

**Comment by pskowronek** [comment]
I've just added additional search options and replace all to internal viewer/editor.
Probably it would be nice to have at least search for binary editor (a feature that is very helpful if binary editor is needed). @ahadas, any thoughts?

**Comment by pskowronek** [comment]
Moved to discussion here: https://github.com/mucommander/mucommander/discussions/914

**Comment by ahadas** [comment]
Version 1.3.0 that includes a fix for this issue [has been released](https://github.com/mucommander/mucommander/releases/tag/1.3.0-1)

**Pull request merged**